js/shared/main/scala/upickle/Annotator.scala

```diff
@@ -0,0 +1,43 @@
+package upickle
+
+import scala.annotation.implicitNotFound
+import scala.reflect.ClassTag
+
+/**
+ * Represents a strategy for storing a type annotation in the serialized data.
+ *
+ * The strategy wraps the given reader and writer to load and store the type annotation. It is supposed
+ * that actual implementations have compatible `annotate` methods.
+ */
+@implicitNotFound("Couldn't find any Annotator implementation in scope. Make sure you have imported some configuration, e.g. `import upickle.config.default._`")
+trait Annotator {
+  def annotate[V](r: Reader[V], name: String): Reader[V]
+  def annotate[V: ClassTag](w: Writer[V], name: String): Writer[V]
+}
+
+object Annotator {
+  // It drops down to array annotations if the serialized object is not represented as a `Js.Obj`
+  def keyAnnotator(key: String): Annotator = new Annotator {
```

Comment:
Ok, let me rephrase: Is there a specific reason why you didn't write: ``case class KeyAnnotator(key: String) extends Annotator { ... }``?